File: src/application/models/lists/Monsters.java
package application.models.lists;

import java.io.Serializable;
import java.util.List;

import application.ROM;
import application.controllers.MainMenu;
import application.models.Monster;

public class Monsters extends Models<Monster> implements Serializable
{
    private static final int monsterListSize = 98;

    public Monsters()
    {
	super();
    }

    public Monsters(Monsters addFrom, boolean keepEmptyModel)
    {
	super(addFrom, keepEmptyModel);
    }

    @Override
    public void initializeModel(List<Monster> models)
    {
	for (int i = 0; i < monsterListSize; i++)
	{
	    Monster monster = new Monster(i);
	    monster.getValuesFromROM();

	    if (ROM.showMonsterDuplicates == false)
	    {

		if (models.contains(monster)) // if monster is duplicate of
					      // existing monster, add it to
					      // that monsters list of aliases
		{
		    Monster monsterInList = models.get(models.indexOf(monster));
		    monsterInList.getAliases().add(monster);
		}

		else
		{
		    models.add(monster);
		}
	    }
	    else
	    {
		models.add(monster);
	    }
	}
    }

    @Override
    public void saveModels()
    {
	for (Monster m : models)
	{
	    m.writeValuesToROM();
	    for (Monster alias : m.getAliases())
	    {
		// System.out.println("alias of "+m.getName()+" is
		// "+alias.getName()+" gameIndexes are "+m.getGameIndex()
		// +" and "+ alias.getGameIndex());
		alias.writeValuesToROM();
	    }
	}
    }

}

File: src/application/models/lists/Monsters.kt
package application.models.lists

import application.ROM
import application.Settings
import application.models.Monster
import java.io.Serializable

class Monsters : Models<Monster>, Serializable
{

	constructor() : super()
	constructor(addFrom: Monsters, keepEmptyModel: Boolean) : super(addFrom, keepEmptyModel)
	constructor(addFrom: Monsters, showEmptyValues: Boolean, showDuplicates: Boolean){addUsefulMonsters(addFrom, showEmptyValues, showDuplicates)}

	override fun initialize(models: ModelsList<Monster>)
	{
		for (i in 0..monsterListSize - 1)
		{
			val monster = Monster(i)
			monster.getValuesFromROM()
			models.add(monster)
		}
	}

	fun addUsefulMonsters(addFrom: Monsters, showEmptyValues: Boolean, showDuplicates: Boolean)
	{
		addFrom.models
				.filter {it.name.trim().isNotEmpty() || showEmptyValues}
				.forEach {

					if(showDuplicates == false)
					{
						if (models.contains(it))
						// if monster is duplicate of existing monster,
						// add it to that monsters list of aliases
						{
							val monsterInList = models[models.indexOf(it)]
							monsterInList.aliases.add(it)
						}
						else models.add(it)
					}
					else
					{models.add(it)}

		}
		dModels.addAll(addFrom.dModels)
	}

	override fun saveModels()
	{
		models.forEach {
			it.writeValuesToROM()
			it.aliases.forEach { it.writeValuesToROM() }
		}
	}

	companion object
	{
		private val monsterListSize = 98
	}

}
